From ba0ff1d15772e1e66bc7c6fc8ba594207d304d1b Mon Sep 17 00:00:00 2001
From: Samuel Dionne-Riel <samuel@dionne-riel.com>
Date: Thu, 18 Mar 2021 13:20:56 -0400
Subject: [PATCH] himax_0flash_mmi: Fix for built-in module

The load order would be wrong. No "IC" modules were present when the
main himax module was started, as the "IC" modules were running after
the main himax module.

Changing the makefile order changes the linking order, which in turn
affects the order in which modules gets initialized.
---
 .../touchscreen/himax_0flash_mmi/Makefile     | 25 ++++++++++---------
 1 file changed, 13 insertions(+), 12 deletions(-)

diff --git a/drivers/input/touchscreen/himax_0flash_mmi/Makefile b/drivers/input/touchscreen/himax_0flash_mmi/Makefile
index bfe92f7b90fc3..869e9b60b6706 100644
--- a/drivers/input/touchscreen/himax_0flash_mmi/Makefile
+++ b/drivers/input/touchscreen/himax_0flash_mmi/Makefile
@@ -41,6 +41,19 @@ ifneq ($(filter m y,$(CONFIG_TOUCHSCREEN_HIMAX_PANEL_NOTIFICATIONS)),)
 	ccflags-y += -DHIMAX_CONFIG_PANEL_NOTIFICATIONS
 endif
 
+himax_0flash_mmi_hx83112f-y += support_ic/himax_0flash_mmi_hx83112f/himax_ic_HX83112f.o
+obj-$(CONFIG_TOUCHSCREEN_HIMAX_IC_HX83112F) += himax_0flash_mmi_hx83112f.o
+
+himax_0flash_mmi_hx83112-y += support_ic/himax_0flash_mmi_hx83112/himax_ic_HX83112.o
+obj-$(CONFIG_TOUCHSCREEN_HIMAX_IC_HX83112) += himax_0flash_mmi_hx83112.o
+
+himax_0flash_mmi_hx83102-y += support_ic/himax_0flash_mmi_hx83102/himax_ic_HX83102.o
+obj-$(CONFIG_TOUCHSCREEN_HIMAX_IC_HX83102) += himax_0flash_mmi_hx83102.o
+
+himax_0flash_mmi_hx83102d-y += support_ic/himax_0flash_mmi_hx83102d/himax_ic_HX83102d.o
+obj-$(CONFIG_TOUCHSCREEN_HIMAX_IC_HX83102D) += himax_0flash_mmi_hx83102d.o
+
+# Must be last, otherwise when built-in it will probe before the `ic` modules.
 ifneq ($(filter m y,$(CONFIG_TOUCHSCREEN_HIMAX_SPI)),)
 	himax_0flash_mmi-y += himax_platform_QCT_SPI.o
 	ccflags-y += -D__HIMAX_PLATFORM_SPI__ \
@@ -53,15 +66,3 @@ himax_0flash_mmi-$(CONFIG_TOUCHSCREEN_HIMAX_INCELL) += himax_ic_incell_core.o
 himax_0flash_mmi-$(CONFIG_TOUCHSCREEN_HIMAX_INSPECT) += himax_inspection.o
 
 obj-$(CONFIG_TOUCHSCREEN_HIMAX_SPI) += himax_0flash_mmi.o
-
-himax_0flash_mmi_hx83112f-y += support_ic/himax_0flash_mmi_hx83112f/himax_ic_HX83112f.o
-obj-$(CONFIG_TOUCHSCREEN_HIMAX_IC_HX83112F) += himax_0flash_mmi_hx83112f.o
-
-himax_0flash_mmi_hx83112-y += support_ic/himax_0flash_mmi_hx83112/himax_ic_HX83112.o
-obj-$(CONFIG_TOUCHSCREEN_HIMAX_IC_HX83112) += himax_0flash_mmi_hx83112.o
-
-himax_0flash_mmi_hx83102-y += support_ic/himax_0flash_mmi_hx83102/himax_ic_HX83102.o
-obj-$(CONFIG_TOUCHSCREEN_HIMAX_IC_HX83102) += himax_0flash_mmi_hx83102.o
-
-himax_0flash_mmi_hx83102d-y += support_ic/himax_0flash_mmi_hx83102d/himax_ic_HX83102d.o
-obj-$(CONFIG_TOUCHSCREEN_HIMAX_IC_HX83102D) += himax_0flash_mmi_hx83102d.o
-- 
2.29.2

